ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PointSchedule.hs

```diff
@@ -312,6 +324,33 @@ mkPeers h as =
 pointSchedule :: [Tick] -> NonEmpty PeerId -> Maybe PointSchedule
 pointSchedule ticks nePeerIds = (`PointSchedule` nePeerIds) <$> nonEmpty ticks
 
+peerStates :: PeerId -> [(DiffTime, SchedulePoint)] -> [(DiffTime, Peer NodeState)]
+peerStates peerId =
+  fmap (second (Peer peerId . NodeOnline)) .
+  scanl' step (0, zero)
+  where
+    step (_, z) =
+      fmap $ \case
+        ScheduleTipPoint tip -> z {tip = TipPoint tip}
+        ScheduleHeaderPoint h -> z {header = HeaderPoint h}
+        ScheduleBlockPoint b -> z {block = BlockPoint b}
+
+    zero = AdvertisedPoints {
+      tip = TipPoint TipGenesis,
+      header = HeaderPoint (TestHeader genesisBlock),
+      block = BlockPoint genesisBlock
+    }
+
+    genesisBlock = (firstBlock 0) {tbSlot = 0}
+
+fromSchedulePoints :: Map PeerId [(DiffTime, SchedulePoint)] -> Maybe PointSchedule
```

Comment:
Could we get rid of `NodeState` and feed the `SchedulePoint`s one by one to the simulated peer?

Reply:
you mean `AdvertisedPoints`? what would be the benefit of that?

Reply:
The advantages would be that we don't need this conversion anymore, and that we don't repeat points redundantly in the schedule.

Reply:
The peer simulator would then have to maintain the unchanged points in a TVar, since we need all points in each tick. I would consider that less desirable than having to create an intermediate type.

Alternatively, we'd have to pattern match on the `SchedulePoint` at each consumer site.

If we wanted to print the schedule in table format, we'd also have to run the same fold…I think having this representation as the input is quite comfortable.

Reply:
mmm, I'm undecided :)

But I think your rationale would be worth documenting next to the `NodeState` type.

> I would consider that less desirable than having to create an intermediate type.

Why is it less desirable?

Reply:
> Why is it less desirable?

my motivation would be to introduce as little mutable concurrent state as possible if it can reasonably be made declarative, and I think this situation qualifies well